### PR TITLE
(wip) added eslint

### LINKED ISSUE #274: Enable ESLint
I’d go with airbnb config. Semicolons ftw 😄 .
If somebody gets to it, might as well enable Travis (#247).
This should be done before 3.0.

Don’t forget the development happens on the `next` branch.

### PR DESCRIPTION
**Pull request by nfcampos**
@gaearon want to review the config?

I still need to fix most of the files to conform to the config

closes #274

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "airbnb",
+  "parserOptions": {
+    "ecmaFeatures": {
+      "experimentalObjectRestSpread": true
+    },
+  },
+  "rules": {
+    "no-underscore-dangle": ["error", { "allow": ["__REACT_HOT_LOADER__"] }],
+    "jsx-quotes": ["error", "prefer-single"]
```
Comment (gaearon) [inline .eslintrc]: I’m fine with using double quotes everywhere.

File: test/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    },
+  },
+  rules: {
+    "react/prefer-stateless-function": 0,
+    "react/no-multi-comp": 0,
+    "react/prefer-es6-class": 0,
+    "react/prop-types": 0,
+    "no-shadow": 0,
+    "new-cap": 0,
+  },
```
Comment (gaearon) [inline test/.eslintrc]: We might also want to disable it completely for Babel test fixtures.

**Comment by gaearon** [comment]
Looks good. I left a few comments.
You can try using `eslint --fix .` to get some of these fixed.

File: .eslintrc (hunk 1 above)
Comment (nfcampos) [inline .eslintrc]: using double quotes for jsx and single quotes for everything else isn't weird?

File: test/.eslintrc (hunk 2 above)
Comment (nfcampos) [inline test/.eslintrc]: I have, that's why I created a `.eslintignore` file

**Comment by nfcampos** [comment]
Yeah --fix helps at least with the semicolons, adding semicolons in hundreds of lines manually would have been terrible

**Comment by nfcampos** [comment]
@gaearon you wrote some of the files without `const`, `import`, stuff like that, any reason to keep them like that, since we're building everything with babel? I need to either change them or change the eslint rules

File: .eslintrc (hunk 1 above)
Comment (gaearon) [inline .eslintrc]: It’s weird but apparently it’s what everyone else is doing these days 😄

File: test/.eslintrc (hunk 2 above)
Comment (gaearon) [inline test/.eslintrc]: Ah, missed that one.

**Comment by gaearon** [comment]
If we use ES modules everywhere, we should add the commonjs compat plugin so that `react-hot-loader/webpack` and  `react-hot-loader/babel` keep working as they are.

**Comment by nfcampos** [comment]
add the plugin to babel?

File: .eslintrc (hunk 1 above)
Comment (nfcampos) [inline .eslintrc]: double quotes it is then 😄

**Comment by gaearon** [comment]
Yeah, this one: https://www.npmjs.com/package/babel-plugin-add-module-exports. Not sure if it works well though. Never tried it.

**Comment by nfcampos** [comment]
hmm, not worth it then I think. basically I'm getting a bunch of `error  'use strict' is unnecessary inside of modules` from this rule http://eslint.org/docs/rules/strict + `airbnb`  config adding `parserOptions: {sourceType: 'module' }` to the eslint config.  So all of the non-es6 modules files in src/ are being interpreted as being modules and thus not needing 'use strict'. but i dont think it's possible to disable parserOptions on a single file. should i just disable the strict rule?

**Comment by gaearon** [comment]
I’m a bit confused. I meant that we can use ES modules everywhere in the code. We just need to make sure that the CommonJS compat code is generated as part of Babel compilation for entry points like `/babel` and `/webpack`. Does this make sense? Or are you saying this makes ESLint fail?

**Comment by nfcampos** [comment]
what i'm saying is that it currently works great the way it's written so instead of changing everything to es6 modules just to appease eslint it's easier to selectively disable the offending rule, no?

**Comment by gaearon** [comment]
>what i'm saying is that it currently works great the way it's written so instead of changing everything to es6 modules just to appease eslint it's easier to selectively disable the offending rule, no?

Works for me 👍

**Comment by nfcampos** [comment]
Alright, I've finished fixing all the linting errors. Give it a look if you want, I think I didn't break anything. Let's merge this soon to avoid merge conflicts

File: src/AppContainer.dev.js [hunk 3]
```diff
@@ -54,38 +54,46 @@ class AppContainer extends Component {
...
       );
     }
+
+    return undefined;
   },
-  children: function (props) {
+  children(props) {
     if (React.Children.count(props.children) !== 1) {
-      return new Error(`Invalid prop "children" supplied to AppContainer. Expected a single React element with your app’s root component, e.g. <App />.`);
+      return new Error('Invalid prop "children" supplied to AppContainer.' +
```
Comment (gaearon) [inline src/AppContainer.dev.js]: There’s a missing space after the period in the error message. The sentences will collide.
Also a style nit: let’s put a newline right after the opening paren.

File: src/AppContainer.dev.js [hunk 4]
```diff
@@ -54,38 +54,46 @@ class AppContainer extends Component {
...
+        'Expected a single React element with your app’s root component, e.g. <App />.');
     }
-  }
+
+    return undefined;
```
Comment (gaearon) [inline src/AppContainer.dev.js]: Lol this rule is so nitpicky. But I can live with it 😄

File: src/webpack/tagCommonJSExports.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
-;(function () {
+/* global __FILENAME__ */
```
Comment (gaearon) [inline src/webpack/tagCommonJSExports.js]: Kinda silly to leak these into the generated code. Can we add some naïve code to cut comments in the Webpack loader?

**Comment by gaearon** [comment]
The error message needs to be fixed but the rest is up to your judgement.
:shipit:

File: src/AppContainer.dev.js (hunk 3 above)
Comment (nfcampos) [inline src/AppContainer.dev.js]: Good catch

File: src/AppContainer.dev.js (hunk 4 above)
Comment (nfcampos) [inline src/AppContainer.dev.js]: Yeah this one looks pretty silly in some places :)

File: src/webpack/tagCommonJSExports.js (hunk 5 above)
Comment (nfcampos) [inline src/webpack/tagCommonJSExports.js]: I'll do that later

**Pull request merged**